objects/src/assets/mod.rs

```diff
@@ -0,0 +1,420 @@
+        let mut result = [0_u8; 32];
+        let id_bytes: [u8; 24] = asset.faucet_id.into();
+        result[..24].copy_from_slice(&id_bytes);
+        result[24..].copy_from_slice(&asset.amount.to_le_bytes());
+        result
+    }
+}
+
+impl From<FungibleAsset> for Asset {
+    fn from(asset: FungibleAsset) -> Self {
+        Asset::Fungible(asset)
+    }
+}
+
+impl TryFrom<Word> for FungibleAsset {
+    type Error = AssetError;
+
+    fn try_from(value: Word) -> Result<Self, Self::Error> {
+        let faucet_id = AccountId::try_from([value[0], value[1], value[2]])
+            .map_err(AssetError::invalid_account_id)?;
+        let amount = value[3].as_int();
+        Self::new(faucet_id, amount)
+    }
+}
+
+impl TryFrom<[u8; 32]> for FungibleAsset {
+    type Error = AssetError;
+
+    fn try_from(value: [u8; 32]) -> Result<Self, Self::Error> {
+        let word = parse_word(value)?;
+        Self::try_from(word)
+    }
+}
+
+impl fmt::Display for FungibleAsset {
+    fn fmt(&self, _f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        todo!()
+    }
+}
+
+// NON-FUNGIBLE ASSET
+// ================================================================================================
+/// A commitment to a non-fungible asset.
+///
+/// A non-fungible asset consists of 4 field elements which are computed by hashing asset data
+/// (which can be of arbitrary length) with the ID of the faucet which issued the asset, and then
+/// setting bits in the result to ensure the the asset encoding is correct.
+///
+/// Specifically:
+/// - The first bit of the asset is set to ZERO.
+/// - The 32 least significant bits of the 3rd element are set to 2^31.
+///
+/// [NonFungibleAsset] itself does not contain the actual asset data. The container for this data
+/// [NonFungibleAssetDetails] struct.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct NonFungibleAsset(Word);
+
+impl NonFungibleAsset {
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns a non-fungible asset created from the specified asset details.
+    ///
+    /// # Errors
+    /// Returns an error if the provided faucet ID is not for a non-fungible asset faucet.
+    pub fn new(details: &NonFungibleAssetDetails) -> Result<Self, AssetError> {
+        let asset_hash = Hasher::hash(details.asset_data().as_bytes());
+        Self::from_raw_parts(details.faucet_id(), asset_hash)
+    }
+
+    /// Return a non-fungible asset created from the specified faucet and using the provided
+    /// hash of the asset's data.
+    ///
+    /// Hahs of the asset's data is expected to be computed from the binary representation of the
+    /// asset's data.
+    ///
+    /// # Errors
+    /// Returns an error if the provided faucet ID is not for a non-fungible asset faucet.
+    pub fn from_raw_parts(faucet_id: AccountId, data_hash: Digest) -> Result<Self, AssetError> {
+        if !faucet_id.is_non_fungible_faucet() {
```

Comment:
`Self::validate` assumes the NFA was created correctly, but shouldn't this check be part of the validate? Under the risk of performing this check twice (once upon creation, again during asset validation)

Reply:
At the time when `Self::validate()` is called we no longer have access to the original `faucet_id()` and thus cannot perform this check.

It might be good to call `Self::validate()` at the end of constructor in debug mode to make sure construction and validation logic don't diverge. I've added that.